File: server/internal/domain/strategy/composite/contract.go
package composite

import (
	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/common"
	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/strategy"
)

type ResolvedSignal = strategy.ResolvedSignal
type CombinationPolicy = strategy.CombinationPolicy
type CompositeDefinition = strategy.CompositeDefinition

type SignalCombiner interface {
	Combine([]ResolvedSignal, CombinationPolicy) (strategy.Signal, error)
}

type WeightedVoteCombiner struct{}
type MajorityVoteCombiner struct{}

func (WeightedVoteCombiner) Combine([]ResolvedSignal, CombinationPolicy) (strategy.Signal, error) {
	return strategy.Signal{}, common.ErrNotImplemented
}

func (MajorityVoteCombiner) Combine([]ResolvedSignal, CombinationPolicy) (strategy.Signal, error) {
	return strategy.Signal{}, common.ErrNotImplemented
}

File: py_backend/app/domain/strategy/composite/contract.py
"""Composite signal combination (stub).

Mirrors `server/internal/domain/strategy/composite/contract.go`.
"""

from __future__ import annotations

from typing import Protocol

from ..contract import CombinationPolicy, ResolvedSignal, Signal


class SignalCombiner(Protocol):
    def combine(self, children: list[ResolvedSignal], policy: CombinationPolicy) -> Signal: ...


class WeightedVoteCombiner:
    def combine(self, children: list[ResolvedSignal], policy: CombinationPolicy) -> Signal:
        raise NotImplementedError


class MajorityVoteCombiner:
    def combine(self, children: list[ResolvedSignal], policy: CombinationPolicy) -> Signal:
        raise NotImplementedError
